### ISSUE TITLE
array_slice prevents row values to be used more than once

### ISSUE THREAD

**Issue by kryonix** [duckdb/duckdb#3005]
#### What happens?

Function `array_slice` prevents table rows to be used more than once.

- `Q1` (see the examples below) works just fine. `p.loc` is used exactly once in `array_slice(arr, NULL, loc)`.
- `Q2` works, too. Again, `p.loc` is used once in the `WHERE`-clause `AND y = loc`.
- `Q3` uses `p.loc` twice, once in `array_slice`, once in the `WHERE`-clause. This fails with:
   `Error: Not implemented Error: FIXME unimplemented vector type for VectorOperations::Copy`
- The error disappears, when `array_slice(arr, NULL, loc)` is replaced with `array[arr[loc]]` in `Q3`.

These tests lead me to believe that `array_slice` is causing this strange behavior.

#### To Reproduce

```sql
CREATE TABLE p(loc int8);
INSERT INTO p VALUES (1);

-- Q1: (works fine)
WITH RECURSIVE t(y, arr) AS
(
  SELECT 1, array[1,2,3,4,5,6]
    UNION ALL
  SELECT y+1, array_slice(arr, NULL, loc)
  FROM   t, p
  WHERE y < 10
--    AND y = loc
) SELECT * FROM t;

-- Q2: (works fine)
WITH RECURSIVE t(y, arr) AS
(
  SELECT 1, array[1,2,3,4,5,6]
    UNION ALL
  SELECT y+1, arr --array_slice(arr, NULL, loc)
  FROM   t, p
  WHERE y < 10
    AND y = loc
) SELECT * FROM t;

-- Q3: (does not work)
WITH RECURSIVE t(y, arr) AS
(
  SELECT 1, array[1,2,3,4,5,6]
    UNION ALL
  SELECT y+1, array_slice(arr, NULL, loc)
  FROM   t, p
  WHERE y < 10
    AND y = loc
) SELECT * FROM t;
```

#### Environment (please complete the following information):
 - OS: macOS Monterey
 - DuckDB Version: git head
 - DuckDB Client: shell

#### Before Submitting

- [x] **Have you tried this on the latest `master` branch?**
* **Python**: `pip install duckdb --upgrade --pre`
* **R**: `install.packages("https://github.com/duckdb/duckdb/releases/download/master-builds/duckdb_r_src.tar.gz", repos = NULL)`
* **Other Platforms**: You can find binaries [here](https://github.com/duckdb/duckdb/releases/tag/master-builds) or compile from source.

- [x] **Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?**

**Comment by l1t1** [comment]
```
D CREATE TABLE p(loc int8);
D INSERT INTO p VALUES (1);
D
D WITH RECURSIVE t(y, arr) AS
> (
> SELECT 1, array[1,2,3,4,5,6]
> UNION ALL
> SELECT y+1, array_slice(arr, NULL, loc)
> FROM   t, p
> WHERE y < 10
>
> ) SELECT * FROM t;
┌────┬────────────────────┐
│ y  │        arr         │
├────┼────────────────────┤
│ 1  │ [1, 2, 3, 4, 5, 6] │
│ 2  │ [1]                │
│ 3  │ [1]                │
│ 4  │ [1]                │
│ 5  │ [1]                │
│ 6  │ [1]                │
│ 7  │ [1]                │
│ 8  │ [1]                │
│ 9  │ [1]                │
│ 10 │ [1]                │
└────┴────────────────────┘
D
D WITH RECURSIVE t(y, arr) AS
> (
> SELECT 1, array[1,2,3,4,5,6]
> UNION ALL
> SELECT y+1, arr --array_slice(arr, NULL, loc)
> FROM   t, p
> WHERE y < 10
> AND y = loc
> ) SELECT * FROM t;
┌───┬────────────────────┐
│ y │        arr         │
├───┼────────────────────┤
│ 1 │ [1, 2, 3, 4, 5, 6] │
│ 2 │ [1, 2, 3, 4, 5, 6] │
└───┴────────────────────┘
D
D WITH RECURSIVE t(y, arr) AS
> (
> SELECT 1, array[1,2,3,4,5,6]
> UNION ALL
> SELECT y+1, array_slice(arr, NULL, loc)
> FROM   t, p
> WHERE y < 10
> AND y = loc
> ) SELECT * FROM t;
┌───┬────────────────────┐
│ y │        arr         │
├───┼────────────────────┤
│ 1 │ [1, 2, 3, 4, 5, 6] │
└───┴────────────────────┘
Error: Not implemented Error: FIXME unimplemented vector type for VectorOperations::Copy
```

**Comment by l1t1** [comment]
```
D WITH RECURSIVE t(y, arr) AS
> (
> SELECT 1, array[1,2,3,4,5,6]
> UNION ALL
> SELECT y+1, array[arr[loc]]
> FROM   t, p
> WHERE y < 10
> AND y = loc
> ) SELECT * FROM t;
┌───┬────────────────────┐
│ y │        arr         │
├───┼────────────────────┤
│ 1 │ [1, 2, 3, 4, 5, 6] │
│ 2 │ [2]                │
└───┴────────────────────┘
D
D WITH RECURSIVE t(y, arr) AS
> (
> SELECT 1, array[1,2,3,4,5,6]
> UNION ALL
> SELECT y+1, arr[loc:6]
> FROM   t, p
> WHERE y < 10
> AND y = loc
> ) SELECT * FROM t;
┌───┬────────────────────┐
│ y │        arr         │
├───┼────────────────────┤
│ 1 │ [1, 2, 3, 4, 5, 6] │
└───┴────────────────────┘
Error: Not implemented Error: FIXME unimplemented vector type for VectorOperations::Copy
```

**Comment by l1t1** [comment]
the i386 0.31version crashed when running q3.

**Comment by l1t1** [comment]
it is related to `RECURSIVE`

both 
```
with t(a,b) as(SELECT 1, array[1,2,3,4,5,6]),
t1(c) as(select * from generate_series(1,6))
select a,c,b[c:6] from t,t1;
```
and 
```
CREATE TABLE p(loc int8);
INSERT INTO p VALUES (1),(2),(3),(4);
with t(a,b) as(SELECT 1, array[1,2,3,4,5,6])
select a,loc,b[loc:6] from t,p;
```
works.
the result of 
```
WITH RECURSIVE t(y, arr) AS
(
  SELECT 1, array[1,2,3,4,5,6]
    UNION ALL
  SELECT y+1, array_slice(arr, y, 6)
  FROM   t
  WHERE y < 6
) SELECT * FROM t;
```
is wrong.

**Comment by l1t1** [comment]
this one column CTE table is also wrong
```
WITH RECURSIVE t(arr) AS
(
  SELECT array[1,2,3,4,5,6]
    UNION ALL
  SELECT  array_slice(arr, arr[0], 6)
  FROM   t
  WHERE arr[0] < 6
) SELECT * FROM t;
┌────────────────────┐
│        arr         │
├────────────────────┤
│ [1, 2, 3, 4, 5, 6] │
│ [2, 3, 4, 5, 6]    │
│                    │
└────────────────────┘
```

**Comment by Mytherin** [comment]
Thanks for the report! I have pushed a fix in #3013.
